### PR TITLE
Silence some follow-up errors [3/x]

### PR DESCRIPTION
**Pull request by oli-obk**
this is one piece of the requested cleanups from https://github.com/rust-lang/rust/pull/117449

Keep error types around, even in obligations.

These help silence follow-up errors, as we now figure out that some types (most notably inference variables) are equal to an error type.

But it also allows figuring out more types in the presence of errors, possibly causing more errors.

### CODE REVIEW

File: tests/ui/impl-trait/issue-72911.rs [hunk 1]
```diff
@@ -15,7 +15,7 @@ fn gather_from_file(dir_entry: &foo::MissingItem) -> impl Iterator<Item = Lint>
 
 fn lint_files() -> impl Iterator<Item = foo::MissingItem> {
     //~^ ERROR: failed to resolve
-    unimplemented!()
+    std::iter::empty()
```
Comment (compiler-errors) [inline tests/ui/impl-trait/issue-72911.rs:18]: why did you change this test?

File: compiler/rustc_infer/src/infer/opaque_types.rs [hunk 2]
```diff
@@ -631,13 +631,6 @@ impl<'tcx> InferCtxt<'tcx> {
                 ct_op: |ct| ct,
             });
 
-            if let ty::ClauseKind::Projection(projection) = predicate.kind().skip_binder() {
-                if projection.term.references_error() {
-                    // No point on adding any obligations since there's a type error involved.
```
Comment (compiler-errors) [inline compiler/rustc_infer/src/infer/opaque_types.rs:636]: Can you point me to the PR that added this originally?

File: tests/ui/impl-trait/issue-72911.rs (hunk 1 above)
Comment (oli-obk) [reply tests/ui/impl-trait/issue-72911.rs:18]: because it otherwise emits a "`(): Iterator` does not hold" error

Comment (compiler-errors) [reply tests/ui/impl-trait/issue-72911.rs:18]: So this PR doesn't end up silencing errors, but also introducing new errors in some cases? 🤔 I think it may be important to note that.

**Review by estebank — APPROVED** [review]

File: tests/ui/impl-trait/issue-72911.rs (hunk 1 above)
Comment (oli-obk) [reply tests/ui/impl-trait/issue-72911.rs:18]: I reverted the change to show the diagnostic change

**Review by compiler-errors — APPROVED** [review]
